crates/sui-core/src/authority_active/checkpoint_executor/mod.rs

```diff
@@ -0,0 +1,252 @@
+                Some(Ok((seq_num, checkpoint))) = finished.next() => {
+                    self.highest_executed_checkpoint = Some(seq_num as CheckpointSequenceNumber);
+                    self.metrics.last_executed_checkpoint.set(seq_num as i64);
+                    self.update_lag_metric(seq_num)
+                        .unwrap_or_else(|err| error!("Update lag metric error: {:?}", err));
+
+                    self.checkpoint_store
+                        .update_highest_executed_checkpoint(&checkpoint)
+                        .unwrap();
+                },
+                // Limits to `task_limit` worker tasks. If we are at capacity, reloop and either
+                // check again or attempt to process and consume the FuturesOrdered queue for
+                // finished tasks
+                received = self.mailbox.recv(), if finished.len() < task_limit => match received {
+                    Ok(checkpoint) => {
+                        let last_to_exec = checkpoint.sequence_number() as u64;
+
+                        // Note that either of these can be higher. If the node crashes with many
+                        // scheduled tasks, then the in-memory watermark starts as None, but the
+                        // persistent watermark is set, hence we start there. If we get a new
+                        // messsage with checkpoints tasks scheduled, then the in-memory watermark
+                        // will be greater, and hence we start from there.
+                        let next_to_exec = std::cmp::max(
+                            self.highest_executed_checkpoint
+                                .map(|highest| highest + 1)
+                                .unwrap_or(0),
+                            highest_scheduled_checkpoint
+                                .map(|highest| highest + 1)
+                                .unwrap_or(0),
+                        );
+
+                        for seq_num in next_to_exec..=last_to_exec {
+                            let next_checkpoint = match self
+                                .checkpoint_store
+                                .get_checkpoint_by_sequence_number(seq_num)
+                            {
+                                Ok(Some(new_checkpoint)) => new_checkpoint,
+                                Ok(None) => {
+                                    self.metrics.checkpoint_exec_errors.inc();
+                                    panic!(
+                                        "Checkpoint {:?} does not exist in checkpoint store",
+                                        seq_num,
+                                    );
+                                }
+                                Err(err) => {
+                                    error!("Failed to fetch checkpoint {:?}: {:?}", seq_num, err);
+                                    self.metrics.checkpoint_exec_errors.inc();
+                                    continue;
+                                }
```

Comment:
I don't think its safe to "continue" here because then we could be into a situation where we fail to schedule checkpoint X but happen to schedule checkpoint X + 1. In fact i'm not even sure if storage read errors from rocksdb are really recoverable.